security/security.c

```diff
@@ -307,6 +340,52 @@ static void __init ordered_lsm_parse(const char *order, const char *origin)
 	kfree(sep);
 }
 
+static void __init
+lsm_init_hook_static_slot(struct security_hook_static_slots *hook)
+{
+	struct security_hook_list *pos;
+	struct security_static_slot *slot;
+	int slot_idx, slot_cnt, first_slot;
+
+	slot_cnt = 0;
+	// todo: race condition if hlist modified in between the two foreach
+	hlist_for_each_entry_rcu (pos, hook->head, list) {
+		slot_cnt++;
+	}
+
+	first_slot = SECURITY_STATIC_SLOT_COUNT - slot_cnt;
+	if (first_slot < 0)
+		panic("%s - No static hook slot remaining to add LSM hook.\n",
+		      __func__);
+
+	if (slot_cnt == 0) {
+		hook->first_slot = INT_MAX;
+		return;
+	}
+
+	slot_idx = first_slot;
+	hlist_for_each_entry_rcu (pos, hook->head, list) {
+		slot = &hook->slots[slot_idx];
+		__static_call_update(slot->key, slot->tramp,
+				     pos->hook.generic_func);
+		slot_idx++;
+	}
+
+	hook->first_slot = first_slot;
+}
+
+static void __init lsm_init_static_slots(void)
+{
+	int i, hook_cnt;
+	struct security_hook_static_slots *hooks =
+		(struct security_hook_static_slots *)&security_list_static_slots;
+	hook_cnt = sizeof(security_list_static_slots) /
+		   sizeof(struct security_hook_static_slots);
```

Comment:
If there's no macro for this already, let's define one.

Reply:
in fact, this whole funcion highlights the awkward duality in the LSM code where `security_list_static_slots` and `security_hook_heads` are sort of an array but actually they're a big structure. It might be better to just sidestep that for now by using the `lsm_hook_defs` X macro trick here

Reply:
I agree that it's awkward. Sometimes this is used for the `security_hook_heads` as well. If we use the X macro, we can probably pass the hlist corresponding as well, and remove it from the `struct security_hook_static_slots`.